openvpn/build/build.py

```diff
@@ -152,131 +281,476 @@ def on_error(func, path, exc_info):
-    shutil.copy('../oqs-openssl-output/openssl/lib/libssl.so.1.0.0', stagepath + '/' + OPENVPN_LINUX_PREFIX + '/lib')
+        # TODO: is there a way to check that the other DLLs in
+        # oqs-openssl-win\x86 (e.g., vcruntime140.dll) have the right version
+        # to work with these openssl DLLs? somehow check that the dependencies
+        # of libeay32.dll and ssleay32.dll are present in the x86 folder.
 
-    os.chdir('..')
+    # Start the x64 build
+    shutil.copytree(OPENSSL_OQS_REPO_DIRNAME, 'openssl-oqs-win-x64')
+    with chdir('openssl-oqs-win-x64'):
+        run_command([
+            'perl', 'Configure', 'VC-WIN64A', 'no-asm',
+            'enable-static-engine'])
 
+        run_command(['ms\\do_win64a.bat'])
+        mycwd = os.getcwd()
+        # Before running nmake, we have to run vcvarsall.bat to set the x64
+        # env vars, in the same shell
+        mycwd = os.getcwd()
+        os.system(
+            VCVARSALL + ' amd64 && cd /d ' + mycwd +
+            ' && nmake -f ms\\ntdll.mak')
+
+        shutil.copy(
+            os.path.join('out32dll', 'libeay32.dll'),
+            os.path.join(STAGE_DIR, 'x64'))
+        shutil.copy(
+            os.path.join('out32dll', 'ssleay32.dll'),
+            os.path.join(STAGE_DIR, 'x64'))
+
+
+def build_oqs_openssl_unix(test_build=False):
+    '''
+    Build openssl for linux/unix
+    '''
+    git_restore(OPENSSL_OQS_REPO_DIRNAME)
+
+    with chdir(OPENSSL_OQS_REPO_DIRNAME):
+        target_system = run_command_capture(
+            ['/bin/sh', OPENSSL_CONFIG_PLATFORM_SCRIPT])
+        target_system = target_system.strip()
+        run_command([
+            'sed', '-i org', '1s|^|#!/usr/bin/env perl\\\n#|', 'Configure'])
+        run_command([
+            './Configure', 'shared',
+            '--prefix={PREFIX}'.format(PREFIX=PREFIX),
+            target_system])
+        run_command(['make'])
+        if test_build:
+            run_command(['make', 'test'])
+        run_command([
+            'make', 'install',
+            'INSTALL_PREFIX={STAGE_DIR}'.format(STAGE_DIR=STAGE_DIR),
+            'INSTALL_TOP="/"'])
+
+
+#
+# OpenVPN
+#
+
+
+def build_openvpn():
+    '''
+    Build openvpn on linux/unix/windows
+    '''
+
+    git_restore(OPENVPN_REPO_DIRNAME)
+
+    with chdir(OPENVPN_REPO_DIRNAME):
+        run_command(['autoreconf', '-i', '-f', '-v'])
+        lib_dir = os.path.abspath(
+            os.path.join(STAGE_DIR, PREFIX[1:], 'lib'))
+        include_dir = os.path.abspath(
+            os.path.join(STAGE_DIR, PREFIX[1:], 'include'))
+
+        if not os.path.exists(lib_dir):
+            die(
+                'Did not find lib_dir: {lib_dir} directory.'.format(
+                    lib_dir=lib_dir))
+        if not os.path.exists(include_dir):
+            die(
+                'Did not find include_dir: {include_dir} directory.'.format(
+                    include_dir=include_dir))
+
```

Comment:
So in the current official version of this script, the entire stage subtree gets created by the script, but now your changes seem to expect that subtree to expect ahead of time. Can you confirm this is the intended behavior and, if so, why you changed this?

Reply:
when openvpn is build, it expect ssl library to be already compiled and placed in the lib_dir/include_dir location. later the script instruct the compiler (-L{lib_dir}) to use it, not doing the test would cause openvpn to find any installed (ubuntu) ssl instead and not complain. (confirm)

why: me changed from manually creating the stage/subdirs to make install with prefix because its the clean and correct way to join libraries with applications. mkdir/cp instead mimics the intentions of the library author and will most likely miss things and make it harder to understand the staged structure.